src/rest_framework_dso/embedding.py

```diff
@@ -77,6 +83,95 @@ def _expand_parse_error(allowed_names, field_name):
         )
 
 
+class ChunkedQuerySetIterator(Iterable[M]):
+    """An optimal strategy to perform ``prefetch_related()`` on large datasets.
+
+    It fetches data from the queryset in chunks,
+    and performs ``prefetch_related()`` behavior on each chunk.
+
+    Django's ``QuerySet.prefetch_related()`` works by loading the whole queryset into memory,
+    and performing an analysis of the related objects to fetch. When working on large datasets,
+    this is very inefficient as more memory is consumed. Instead, ``QuerySet.iterator()``
+    is preferred here as it returns instances while reading them. Nothing is stored in memory.
+    Hence, both approaches are fundamentally incompatible. This class performs a
+    mixed strategy: load a chunk, and perform prefetches for that particular batch.
+
+    As extra performance benefit, a local cache avoids prefetching the same records
+    again when the next chunk is analysed. It has a "least recently used" cache to avoid
+    flooding the caches when foreign keys constantly point to different unique objects.
+    """
+
+    def __init__(self, queryset: models.QuerySet, chunk_size=None, sql_chunk_size=None):
+        """
+        :param queryset: The queryset to iterate over, that has ``prefetch_related()`` data.
+        :param chunk_size: The size of each segment to analyse in-memory for related objects.
+        :param sql_chunk_size: The size of each segment to fetch from the database,
+            used when server-side cursors are not available. The default follows Django behavior.
+        """
+        self.queryset = queryset
+        self.sql_chunk_size = sql_chunk_size or DEFAULT_SQL_CHUNK_SIZE
+        self.chunk_size = chunk_size or self.sql_chunk_size
+        self._fk_caches = defaultdict(lambda: LRU(self.chunk_size // 2))
+
+    def __iter__(self):
+        # Using iter() ensures the ModelIterable is resumed with the next chunk.
+        qs_iter = iter(self.queryset.iterator(chunk_size=self.sql_chunk_size))
+
+        # Keep fetching chunks
+        while instances := list(islice(qs_iter, self.chunk_size)):
+            # Perform prefetches on this chunk:
+            if self.queryset._prefetch_related_lookups:
```

Comment:
Maybe a bit of a risk, to rely on internal Django variables? Otoh, upgrading Django is not automatic, so we will detect it when this changes.

Reply:
I doubt this will change that fast, and as last resort we could even remove it if this becomes unportable.
The unit test does cover this case specifically.